### PR TITLE
feat(api,ui): add workflow retention policy

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14589.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14617.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14618.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14621.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘  
  * Standalone tests 1 -   
  * Standalone tests 2 -   
  * Standalone tests 3 -

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14626.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✘  
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14630.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Review by yesnault — APPROVED** [review]

File: cli/cdsctl/admin_workflow.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Short:   "Manage CDS workflows",
+}
+
+func adminWorkflows() *cobra.Command {
+	return cli.NewCommand(adminWorkflowsCmd, nil, []*cobra.Command{
+		cli.NewCommand(adminWorkflowUpdateMaxRunCmd, adminWorkflowUpdateMaxRun, nil),
+	})
+}
+
+// list command
```
Comment (yesnault) [inline cli/cdsctl/admin_workflow.go:20]: the maxrun cmd does not list run

File: docs/content/docs/concepts/files/workflow-syntax.md [hunk 2]
```diff
@@ -35,6 +35,7 @@ notifications:
     on_success: never
     recipients:
     - me@foo.bar
+retention_policy: return run_date_before < 7
```
Comment (yesnault) [inline docs/content/docs/concepts/files/workflow-syntax.md:38]: run_date_before -> run_days_before to be more understandable?

File: docs/content/docs/concepts/workflow/retention.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+---
+title: "Retention"
+weight: 10
+---
+
+Information about workflow execution retention can be found in the workflow advanced section in the CDS UI
```
Comment (yesnault) [inline docs/content/docs/concepts/workflow/retention.md:6]:
```suggestion
You can configure the workflow execution retention in the workflow advanced section on the CDS UI.
```

File: docs/content/docs/concepts/workflow/retention.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+![retention.png](../images/workflow_retention.png)
+
+
+* The first line, define the number maximum of execution that CDS can keep for this workflow. Only a CDS administrator can update this value
```
Comment (yesnault) [inline docs/content/docs/concepts/workflow/retention.md:11]:
```suggestion
* The first line defines the number maximum of executions that CDS can keep for this workflow. Only a CDS administrator can update this value.
```

File: docs/content/docs/concepts/workflow/retention.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+* On the second line, you will be able to define your retention policy through a lua condition.
+You will be able to use these variables:
+  * run_date_before: To identify execution older than x days
```
Comment (yesnault) [inline docs/content/docs/concepts/workflow/retention.md:15]: run_days_before ?

File: docs/content/docs/concepts/workflow/retention.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  * git_branch_exist: To identify if the execution git branch still exist on repository side
```
Comment (yesnault) [inline docs/content/docs/concepts/workflow/retention.md:16]:
```suggestion
  * git_branch_exist: to identify if the git branch used for this execution still exists on the git repository
```

File: docs/content/docs/concepts/workflow/retention.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  * run_status: To identidy execution status
```
Comment (yesnault) [inline docs/content/docs/concepts/workflow/retention.md:17]:
```suggestion
  * run_status: to identify execution status
```

File: docs/content/docs/concepts/workflow/retention.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  * and all variables defined in your workflow payload
+
+For example, the rule defined above means:
+
+Keep workflow execution for 365 days, but if branch does not exist on repository, only keep the execution 2 days.
+ 
+
+* The dry run button allow you to test your lua expression. The result is a table filled with all executions that would be kept
```
Comment (yesnault) [inline docs/content/docs/concepts/workflow/retention.md:25]:
```suggestion
* The dry run button allows you to test your lua expression. The result is a table filled with all the runs that would be kept.
```

File: engine/api/api.go [hunk 9]
```diff
@@ -201,6 +201,9 @@ type Configuration struct {
 		Content string `toml:"content" comment:"Help Content. Warning: this message could be view by anonymous user. Markdown accepted." json:"content" default:""`
 		Error   string `toml:"error" comment:"Help displayed to user on each error. Warning: this message could be view by anonymous user. Markdown accepted." json:"error" default:""`
 	} `toml:"help" comment:"######################\n 'Help' informations \n######################" json:"help"`
+	Workflow struct {
+		MaxRuns int64 `toml:"maxRuns" comment:"Maximum of execution by workflow" json:"maxRuns" default:"255"`
```
Comment (yesnault) [inline engine/api/api.go:205]:
```suggestion
		MaxRuns int64 `toml:"maxRuns" comment:"Maximum of runs by workflow" json:"maxRuns" default:"255"`
```

File: engine/api/api.go [hunk 10]
```diff
@@ -201,6 +201,9 @@ type Configuration struct {
...
+	} `toml:"workflow" comment:"######################\n 'Workflow' globla configuration \n######################" json:"workflow"`
```
Comment (yesnault) [inline engine/api/api.go:206]:
```suggestion
	} `toml:"workflow" comment:"######################\n 'Workflow' global configuration \n######################" json:"workflow"`
```

**Review by yesnault — CHANGES_REQUESTED** [review]

File: engine/api/workflow/init.go [hunk 11]
```diff
@@ -11,12 +11,14 @@ import (
 )
 
 var baseUIURL, defaultOS, defaultArch string
+var maxRuns int64
 
 //Initialize starts goroutines for workflows
-func Initialize(ctx context.Context, DBFunc func() *gorp.DbMap, store cache.Store, uiURL, confDefaultOS, confDefaultArch string, maxLogSize int64) {
+func Initialize(ctx context.Context, DBFunc func() *gorp.DbMap, store cache.Store, uiURL, confDefaultOS, confDefaultArch string, maxLogSize int64, runMax int64) {
```
Comment (yesnault) [inline engine/api/workflow/init.go:17]:
```suggestion
func Initialize(ctx context.Context, DBFunc func() *gorp.DbMap, store cache.Store, uiURL, confDefaultOS, confDefaultArch string, maxLogSize int64, confMaxRuns int64) {
```

File: engine/api/workflow/init.go [hunk 12]
```diff
@@ -11,12 +11,14 @@ import (
...
 	baseUIURL = uiURL
 	defaultOS = confDefaultOS
 	defaultArch = confDefaultArch
+	maxRuns = runMax
```
Comment (yesnault) [inline engine/api/workflow/init.go:21]:
```suggestion
	maxRuns = confMaxRuns
```

File: sdk/messages.go [hunk 13]
```diff
@@ -89,6 +89,7 @@ var (
...
 	MsgWorkflowDeprecatedVersion            = &Message{"MsgWorkflowDeprecatedVersion", trad{FR: "La configuration yaml de votre workflow est dans un format déprécié. Exportez le avec la CLI `cdsctl workflow export %s %s`", EN: "The yaml workflow configuration format is deprecated. Export your workflow with CLI `cdsctl workflow export %s %s`"}, nil, RunInfoTypeWarning}
 	MsgWorkflowGeneratedFromTemplateVersion = &Message{"MsgWorkflowGeneratedFromTemplateVersion", trad{FR: "Le workflow a été généré à partir du modèle de workflow: %s.", EN: "The workflow was generated from the template: %s"}, nil, RunInfoTypInfo}
+	MsgTooMuchWorkflowRun                   = &Message{"MsgTooMuchWorkflowRun", trad{FR: "L'éxécution de ce workflow est suspendu. Vous dépassez le nombre maximum d'éxécution autorisé (%.f). Merci de revoir la politique de retention de ce workflow", EN: "Workflow execution is delayed. The maximum number of executions for this workflow has been reached ( %.f ). Please update your workflow retention policy"}, nil, RunInfoTypeWarning}
```
Comment (yesnault) [inline sdk/messages.go:92]:
```suggestion
	MsgTooMuchWorkflowRun                   = &Message{"MsgTooMuchWorkflowRun", trad{FR: "L'éxecution de ce workflow est suspendu. Vous dépassez le nombre maximum de runs autorisés (%.f). Merci de revoir la politique de retention de ce workflow", EN: "Workflow execution is delayed. The maximum number of runs for this workflow has been reached ( %.f ). Please update your workflow retention policy"}, nil, RunInfoTypeWarning}
```

File: ui/src/assets/i18n/en.json [hunk 14]
```diff
@@ -908,6 +908,12 @@
   "workflow_notification_vcs_comment_always": "Always send",
   "workflow_notification_vcs_pr_comment_body": "Pull-request's comment body",
   "workflow_notification_explanation": "_A user notification can be useful to report the status of a workflow according to its status. Each pipeline in a workflow can be notified based on status in 'Success', 'Fail' or status change. The message sent to the recipients can be set using [CDS variables] (https://ovh.github.io/cds/docs/concepts/variables/). E-mail notifications can also contain HTML, cf. [User Notifications] documentation (https://ovh.github.io/cds/docs/concepts/workflow/notifications/) ._",
+  "workflow_retention_maxruns": "Maximum number of workflow executions: ",
```
Comment (yesnault) [inline ui/src/assets/i18n/en.json:911]:
```suggestion
  "workflow_retention_maxruns": "Maximum number of workflow runs: ",
```

File: ui/src/assets/i18n/en.json [hunk 15]
```diff
@@ -908,6 +908,12 @@
...
+  "workflow_retention_maxruns_admin": "You can contact a CDS administrator to customize this value",
+  "workflow_retention_policy": "Workflow execution retention policy",
+  "workflow_retention_variables": "Available variables: {{vars}}",
+  "workflow_retention_result_title": "The following executions will be kept",
```
Comment (yesnault) [inline ui/src/assets/i18n/en.json:915]:
```suggestion
  "workflow_retention_result_title": "The following runs will be kept",
```

File: ui/src/assets/i18n/en.json [hunk 16]
```diff
@@ -908,6 +908,12 @@
...
+  "workflow_run_delayed": "Workflow execution has been delayed",
```
Comment (yesnault) [inline ui/src/assets/i18n/en.json:916]:
```suggestion
  "workflow_run_delayed": "Workflow run has been delayed",
```

File: ui/src/assets/i18n/fr.json [hunk 17]
```diff
@@ -902,11 +901,18 @@
   "workflow_permission_form_title": "Ajouter une permission sur le workflow",
   "workflow_permission_list_title": "Liste des permissions sur le workflow",
   "workflow_preview_mode": "Votre workflow est dans un état de prévisualisation",
+  "workflow_purged_dry_run_empty": "Aucune donnée ne sera supprimé",
   "workflow_repository_help_line_1": "Votre workflow n'a pas été importé depuis votre code.",
   "workflow_repository_help_line_2": "Gérez votre workflow \"as code\" depuis votre gestionnaire de dépôt pour modifier automatiquement celui-ci avec les changements sur vos branches.",
   "workflow_resync_vcs_tooltip": "Cette action va renvoyer les statuts de vos pipelines à votre gestionnaire de dépots distants",
   "workflow_resync_vcs": "Resynchroniser les statuts VCS",
   "workflow_resync": "Resynchroniser le workflow",
+  "workflow_retention_maxruns": "Nombre maximum d'exécutions du workflow: ",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json:910]:
```suggestion
  "workflow_retention_maxruns": "Nombre maximum de runs du workflow: ",
```

File: ui/src/assets/i18n/fr.json [hunk 18]
```diff
@@ -902,11 +901,18 @@
...
+  "workflow_retention_maxruns_admin": "Vous pouvez contacter un administrateur pour changer cette valeur",
+  "workflow_retention_policy": "Politique de rétention des éxécutions du workflow",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json:912]:
```suggestion
  "workflow_retention_policy": "Politique de rétention des runs du workflow",
```

File: ui/src/assets/i18n/fr.json [hunk 19]
```diff
@@ -902,11 +901,18 @@
...
+  "workflow_retention_variables": "Variables disponibles: {{vars}}",
+  "workflow_retention_result_title": "Les exécutions suivantes seront gardées",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json:914]:
```suggestion
  "workflow_retention_result_title": "Les runs suivants seront gardés",
```

File: ui/src/assets/i18n/fr.json [hunk 20]
```diff
@@ -902,11 +901,18 @@
...
+  "workflow_run_delayed": "L'exécution du workflow a été reportée",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json:915]:
```suggestion
  "workflow_run_delayed": "Le run du workflow a été reporté",
```

File: ui/src/assets/i18n/fr.json [hunk 21]
```diff
@@ -902,11 +901,18 @@
   "workflow_permission_form_title": "Ajouter une permission sur le workflow",
   "workflow_permission_list_title": "Liste des permissions sur le workflow",
   "workflow_preview_mode": "Votre workflow est dans un état de prévisualisation",
+  "workflow_purged_dry_run_empty": "Aucune donnée ne sera supprimé",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json:904]:
```suggestion
  "workflow_purged_dry_run_empty": "Aucune donnée ne sera supprimée",
```

**Review by yesnault — CHANGES_REQUESTED** [review]

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14639.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/api/purge/purge_run.go [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+		if count > offset+limit {
+			offset += limit
+			continue
+		}
+		break
+	}
+	return runsTokeep, nil
+}
+
+func applyRetentionPolicyOnRun(db *gorp.DbMap, wf sdk.Workflow, run sdk.WorkflowRun, branchesMap map[string]struct{}, opts MarkAsDeleteOptions) (bool, error) {
```
Comment (fsamin) [inline engine/api/purge/purge_run.go:98]: i don't really understand what the returned boolean means

File: engine/api/workflow/dao.go [hunk 23]
```diff
@@ -288,14 +293,16 @@ func Insert(ctx context.Context, db gorpmapper.SqlExecutorWithTx, store cache.St
 	if w.HistoryLength == 0 {
 		w.HistoryLength = sdk.DefaultHistoryLength
 	}
+	w.MaxRuns = maxRuns
+	w.RetentionPolicy = "return (git_branch_exist == \"false\" and run_days_before < 2) or run_days_before < 365"
```
Comment (fsamin) [inline engine/api/workflow/dao.go:297]: i wonder is this rule should be in the configuration

File: engine/sql/api/218_workflow_retention_policy.sql [hunk 24]
```diff
@@ -0,0 +1,12 @@
+-- +migrate Up
+alter table "workflow" add column retention_policy TEXT;
+update workflow set retention_policy = 'return (git_branch_exist == "false" and run_days_before < 2) or run_days_before < 365';
```
Comment (fsamin) [inline engine/sql/api/218_workflow_retention_policy.sql:3]: I wonder if we should keep NULL as default value and refers to the default rule stored in API configuration to be able to change the default rule for ALL the workflow easily

**Review by fsamin — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#14640.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/api/purge/purge_run.go (hunk 22 above)
Comment (sguiheux) [reply engine/api/purge/purge_run.go:98]: it's the result of retention policy. True means run is kept.

File: engine/api/workflow_run_craft.go [hunk 25]
```diff
@@ -107,6 +112,50 @@ func (api *API) workflowRunCraft(ctx context.Context, id int64) error {
...
+				Type: sdk.MsgTooMuchWorkflowRun.Type,
+				Args: []interface{}{wf.MaxRuns},
+			}
+			workflow.AddWorkflowRunInfo(run, info)
+			if err := workflow.UpdateWorkflowRun(ctx, api.mustDB(), run); err != nil {
+				log.Error(ctx, "unable to star run %d/%d for workflow %s: %v", run.ID, run.Number, wf.Name, err)
+			}
+			event.PublishWorkflowRun(ctx, *run, wf.ProjectKey)
+			return nil
+		} else {
```
Comment (fsamin) [inline engine/api/workflow_run_craft.go:140]: you can remove the else

File: engine/api/workflow_run_craft.go [hunk 26]
```diff
@@ -107,6 +112,50 @@ func (api *API) workflowRunCraft(ctx context.Context, id int64) error {
...
+			info := sdk.SpawnMsg{
+				ID:   sdk.MsgTooMuchWorkflowRun.ID,
+				Type: sdk.MsgTooMuchWorkflowRun.Type,
+				Args: []interface{}{wf.MaxRuns},
+			}
+			workflow.AddWorkflowRunInfo(run, info)
+			if err := workflow.UpdateWorkflowRun(ctx, api.mustDB(), run); err != nil {
+				log.Error(ctx, "unable to star run %d/%d for workflow %s: %v", run.ID, run.Number, wf.Name, err)
+			}
+			event.PublishWorkflowRun(ctx, *run, wf.ProjectKey)
```
Comment (fsamin) [inline engine/api/workflow_run_craft.go:138]: if there is an error, maybe we shouldn't publish the event ?

File: engine/api/workflow_run_craft.go [hunk 27]
```diff
@@ -107,6 +112,50 @@ func (api *API) workflowRunCraft(ctx context.Context, id int64) error {
...
+					run.Infos[i].Message.Type = sdk.RunInfoTypInfo
+					found = true
+					break
+				}
+			}
+			if found {
+				if err := workflow.UpdateWorkflowRun(ctx, api.mustDB(), run); err != nil {
+					log.Error(ctx, "unable to star run %d/%d for workflow %s: %v", run.ID, run.Number, wf.Name, err)
+				}
+				event.PublishWorkflowRun(ctx, *run, wf.ProjectKey)
```
Comment (fsamin) [inline engine/api/workflow_run_craft.go:154]: if there is an error, maybe we shouldn't publish the event ?

**Review by fsamin — CHANGES_REQUESTED** [review]

File: engine/api/workflow_run_craft.go (hunk 26 above)
Comment (sguiheux) [reply engine/api/workflow_run_craft.go:138]: must not log error but return it

File: engine/api/workflow_run_craft.go (hunk 27 above)
Comment (sguiheux) [reply engine/api/workflow_run_craft.go:154]: must not log error but return it

**Review by fsamin — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

**Pull request merged**